File: src/tasks/ZigZag/ZigZag.hs
import Data.Array (array, bounds, range, (!))
import Text.Printf (printf)
import Data.Monoid (mappend)
import Data.List (sortBy)
 
compZig (x,y) (x',y') =           compare (x+y) (x'+y')
                        `mappend` if even (x+y) then compare x x'
                                                else compare y y'
 
zigZag upper = array b $ zip (sortBy compZig (range b))
                             [0..]
  where b = ((0,0),upper)
 
-- format a 2d array of integers neatly
show2d a = unlines [unwords [printf "%3d" (a ! (x,y) :: Integer) | x <- axis fst] | y <- axis snd]
  where (l, h) = bounds a
        axis f = [f l .. f h]
 
main = mapM_ (putStr . show2d . zigZag) [(3,3), (4,4), (10,2)]

File: src/tasks/ZigZag/ZigZag.java


public class ZigZag {
	
  public static int[][] zigZag(final int size){
    int[][] data = new int[size][size];
    int i = 1;
    int j = 1;

    for (int element = 0; element < size * size; element++) {
      data[i - 1][j - 1] = element;
        
      if ((i + j) % 2 == 0) {
        // Even stripes
        if (j < size)
          j++;
        else
          i += 2;
     
    if (i > 1)
        i--;
    } else {
      // Odd stripes
      if (i < size)
        i++;
      else
        j+= 2;
      if (j > 1)
        j--;
    }
    
    
    }
    return data;
  }

  public static void main(String[] args) {
    int[][] zigzag = ZigZag.zigZag(5);
    for (int i = 0; i < zigzag.length; i++ ) {
        for (int k = 0; k < zigzag[i].length; k++) {
            System.out.print(zigzag[i][k] + ",");
        }
        System.out.println();
    }  
  }
}
